### PR TITLE
pgnodemx Exporter Queries For cgroups v2 Issue: [sc-15818]

### PR DESCRIPTION
**Pull request by tony-landreth**
# Description  

As cgroup v2 support/adoption within Kubernetes continues to progress, users are encountering issues with broken pgnodemx exporter queries. This commit updates the paths where needed and returns 0 values for statistics that are no longer available.

Fixes #  
Version 2 of cgroup changes the names and locations for some of its statistics. For other v1 statistics, there is no exact v2 counterpart, such as memory.stat.rss, memory.stat.mapped_file, memory.kmem.usage_in_bytes.

References:

- [Exporter Queries Incompatible with 'Unified' Cgroup Mode #273](https://github.com/CrunchyData/pgmonitor/issues/273)
- [cgroup v1](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v1/)
- [cgroup v2](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html)
- [Enabling cgroup v2 on RHEL8](https://access.redhat.com/solutions/6898151)
- [cpu.cfs_period_us vs cpu.max](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/managing_monitoring_and_updating_the_kernel/assembly_using-cgroupfs-to-manually-manage-cgroups_managing-monitoring-and-updating-the-kernel)

Depends on #  
An environment where the user has control over the cgroup version. I used GCE VMs and tested against k3s.

## Type of change  
Please check all options that are relevant  
- [x] Bug fix (change which fixes an issue)  
- [ ] Feature (change which adds functionality)  
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)  
- [ ] Documentation update only  

# How Has This Been Tested?  
**Tested Configuration**:  
- [ ] Ansible, Specify version(s):  
- Installation method:  
    - [ ] Binary install from source, version:  
    - [ ] OS package repository, distro, and version:  
    - [ ] Local package server, version:  
    - [ ] Custom-built package, version:  
    - [x] Other:  I used the make targets for postgres-operator.
- [x] PostgreSQL, Specify version(s): 14.4-0  
- [ ] docs tested with hugo version(s):  

Tested with playbook(s):  

# Checklist:  
- [ ] My changes generate no new lint warnings  
- I have made corresponding changes to:  
    - [ ] the documentation  
    - [ ] the release notes  
    - [ ] the upgrade doc

### CODE REVIEW

**Review by keithf4 — DISMISSED** [review]
Code review looks good. Will wait on testing from container team to fully merge to development

**Review by hunleyd — DISMISSED** [review]

File: postgres_exporter/common/queries_nodemx.yml [hunk 1]
```diff
@@ -101,8 +111,19 @@ ccp_nodemx_cpu:
...
 
 ccp_nodemx_cpucfs:
-  query: "SELECT  monitor.cgroup_scalar_bigint('cpu.cfs_period_us') as period_us, 
-                  case when monitor.cgroup_scalar_bigint('cpu.cfs_quota_us') < 0 then 0 else monitor.cgroup_scalar_bigint('cpu.cfs_quota_us') end as quota_us"
+  query: "SELECT
+  CASE
+  WHEN monitor.cgroup_mode() = 'legacy' THEN
+    monitor.cgroup_scalar_bigint('cpu.cfs_period_us')
+  ELSE
+    (SELECT cgroup_array_bigint[2] FROM monitor.cgroup_array_bigint('cpu.max'))
```
Comment (cbandy) [inline postgres_exporter/common/queries_nodemx.yml:119]: 🤔 Can this SELECT go away?

```suggestion
    monitor.cgroup_array_bigint('cpu.max')[2]
```

File: postgres_exporter/common/queries_nodemx.yml [hunk 2]
```diff
@@ -101,8 +111,19 @@ ccp_nodemx_cpu:
...
+  END AS period_us, 
+  CASE
+  WHEN monitor.cgroup_mode() = 'legacy' THEN
+    (CASE WHEN monitor.cgroup_scalar_bigint('cpu.cfs_quota_us') < 0 THEN 0 ELSE monitor.cgroup_scalar_bigint('cpu.cfs_quota_us') END)
+  ELSE
+    (CASE WHEN (SELECT cgroup_array_bigint[1] FROM monitor.cgroup_array_bigint('cpu.max')) < 0 THEN 0 ELSE (SELECT cgroup_array_bigint[1] FROM monitor.cgroup_array_bigint('cpu.max')) END)
```
Comment (cbandy) [inline postgres_exporter/common/queries_nodemx.yml:125]: 🤔 These can be shorter with [GREATEST](https://www.postgresql.org/docs/current/functions-conditional.html#FUNCTIONS-GREATEST-LEAST), I think. (The latter is only correct if SELECT can be eliminated.)

```suggestion
    GREATEST(monitor.cgroup_scalar_bigint('cpu.cfs_quota_us'), 0)
  ELSE
    GREATEST(monitor.cgroup_array_bigint('cpu.max')[1], 0)
```

File: postgres_exporter/common/queries_nodemx.yml [hunk 3]
```diff
@@ -30,24 +30,34 @@ ccp_nodemx_process:
...
- (select val from d where key='inactive_file') as inactive_file,
- monitor.cgroup_scalar_bigint('memory.usage_in_bytes') as usage_in_bytes,
- monitor.cgroup_scalar_bigint('memory.kmem.usage_in_bytes') as kmem_usage_in_bytes
-"
+  monitor.kdapi_scalar_bigint('mem_request') AS request,
+  CASE
+    WHEN monitor.cgroup_mode() = 'legacy' THEN
+        (CASE WHEN monitor.cgroup_scalar_bigint('memory.limit_in_bytes') = 9223372036854771712 THEN 0 ELSE monitor.cgroup_scalar_bigint('memory.limit_in_bytes') END)
+    ELSE
+        (CASE WHEN monitor.cgroup_scalar_bigint('memory.max') = 9223372036854775807 THEN 0 ELSE monitor.cgroup_scalar_bigint('memory.max') END)
```
Comment (cbandy) [inline postgres_exporter/common/queries_nodemx.yml:40]: 🤔 The following is shorter but less clear maybe.

❓ These very-big numbers are not the same. Why's that? I assumed they were MAX_INT or something.

```suggestion
        COALESCE(NULLIF(monitor.cgroup_scalar_bigint('memory.limit_in_bytes'), 9223372036854771712), 0)
    ELSE
        COALESCE(NULLIF(monitor.cgroup_scalar_bigint('memory.max'), 9223372036854775807), 0)
```

File: postgres_exporter/common/queries_nodemx.yml [hunk 4]
```diff
@@ -112,7 +133,11 @@ ccp_nodemx_cpucfs:
         description: "the length of a period (in microseconds)"
 
 ccp_nodemx_cpuacct:
-  query: "SELECT monitor.cgroup_scalar_bigint('cpuacct.usage') as usage, clock_timestamp() as usage_ts"
+  query: "SELECT CASE WHEN monitor.cgroup_mode() = 'legacy'
```
Comment (cbandy) [inline postgres_exporter/common/queries_nodemx.yml:136]: 🌱 I'm curious why this file is using YAML flow and double quotes `"` for these multi-line queries. The block style with `|` or `|-` seems better suited. 🤔 Even for single-line queries, we wouldn't have to worry about escape sequences.

https://yaml-multiline.info/

File: postgres_exporter/common/queries_nodemx.yml [hunk 5]
```diff
@@ -30,24 +30,34 @@ ccp_nodemx_process:
...
+  CASE
+    WHEN monitor.cgroup_mode() = 'legacy'
+      THEN monitor.cgroup_scalar_bigint('memory.usage_in_bytes')
+      ELSE monitor.cgroup_scalar_bigint('memory.current')
+  END as usage_in_bytes,
+  CASE
+    WHEN monitor.cgroup_mode() = 'legacy'
+    THEN monitor.cgroup_scalar_bigint('memory.kmem.usage_in_bytes') 
+    ELSE 0
+  END as kmem_usage_in_byte"
```
Comment (cbandy) [inline postgres_exporter/common/queries_nodemx.yml:60]: ⛏️ Indentation inside these two blocks differs. I prefer the latter here. 🤷🏻

**Review by cbandy — COMMENTED** [review]
Looking good. One question about the differing numeric constants.

File: postgres_exporter/common/queries_nodemx.yml (hunk 3 above)
Comment (tony-landreth) [reply postgres_exporter/common/queries_nodemx.yml:40]: I should find out the why. But, yeah, it looks like the default max differs between v1 and v2.

File: postgres_exporter/common/queries_nodemx.yml (hunk 1 above)
Comment (tony-landreth) [reply postgres_exporter/common/queries_nodemx.yml:119]: It does look funny. The SELECT returns an array named `cgroup_array_bigint`. If you omit the SELECT, you get an error:
```
postgres=# monitor.cgroup_array_bigint('cpu.max');
ERROR:  syntax error at or near "monitor"
LINE 1: monitor.cgroup_array_bigint('cpu.max');
```
If you try to index on the SELECT, you also get an error:
```
postgres=# SELECT monitor.cgroup_array_bigint('cpu.max')[2];
ERROR:  syntax error at or near "["
LINE 1: SELECT monitor.cgroup_array_bigint('cpu.max')[2];
```

File: postgres_exporter/common/queries_nodemx.yml (hunk 2 above)
Comment (tony-landreth) [reply postgres_exporter/common/queries_nodemx.yml:125]: Nice! I'll use GREATEST.

File: postgres_exporter/common/queries_nodemx.yml (hunk 3 above)
Comment (tony-landreth) [reply postgres_exporter/common/queries_nodemx.yml:40]: I still don't know they _why_, but the _what_ looks like cgroup v1 used [the maximum 64-bit signed integer, rounded to the nearest page](https://unix.stackexchange.com/questions/420906/what-is-the-value-for-the-cgroups-limit-in-bytes-if-the-memory-is-not-restricte), whereas v2 is using MAX_INT.

Comment (tony-landreth) [reply postgres_exporter/common/queries_nodemx.yml:40]: On the COALESCE vs CASE refactor -- I like your suggestion. My suspicion is that the CASE statements will be more accessible, but it depends on who's reading :man_shrugging:

File: postgres_exporter/common/queries_nodemx.yml (hunk 1 above)
Comment (cbandy) [reply postgres_exporter/common/queries_nodemx.yml:119]: Looks like it just needs more parentheses.

```sql
CREATE FUNCTION cbandy(TEXT) RETURNS BIGINT[]
LANGUAGE SQL STABLE STRICT
AS $$
  SELECT ARRAY[1000,2000,3000]
$$;

# SELECT cbandy('');
      cbandy
------------------
 {1000,2000,3000}
(1 row)

# SELECT cbandy('')[1];
ERROR:  syntax error at or near "["
LINE 1: SELECT cbandy('')[1];
                         ^
# SELECT (cbandy(''))[1];
 cbandy
--------
   1000
(1 row)

# SELECT (cbandy(''))[2];
 cbandy
--------
   2000
(1 row)
```

**Review by andrewlecuyer — APPROVED** [review]
Tested this on a cluster with cgroup v2 only and these updates appear to be working well.

**Review by cbandy — APPROVED** [review]

File: postgres_exporter/common/queries_nodemx.yml (hunk 1 above)
Comment (tony-landreth) [reply postgres_exporter/common/queries_nodemx.yml]: Nice! I've made the update.

**Review by cbandy — APPROVED** [review]

**Comment by tony-landreth** [comment]
@keithf4 @hunleyd Just letting you know the refactoring has settled down on this PR.

**Comment by andrewlecuyer** [comment]
@keithf4 just wanted to let you know that we've finished reviewing and testing this change, and are OK with merging.

Thanks!!

**Review by keithf4 — APPROVED** [review]
Approving based on feedback from container team. Thanks!

**Review by hunleyd — APPROVED** [review]

**Review by andrewlecuyer — APPROVED** [review]

**Review by keithf4 — APPROVED** [review]

File: postgres_exporter/common/queries_nodemx.yml [hunk 6]
```diff
@@ -30,24 +30,50 @@ ccp_nodemx_process:
...
+  CASE
+    WHEN monitor.cgroup_mode() = 'legacy'
+    THEN (SELECT val FROM d WHERE key='cache')
+    ELSE 0
+  END as cache,
+  CASE
+    WHEN monitor.cgroup_mode() = 'legacy'
+    THEN (SELECT val FROM d WHERE key='rss')
+    ELSE 0
+  END as RSS,
```
[the remark this answers is not in the document]
Comment (keithf4) [reply postgres_exporter/common/queries_nodemx.yml:51]: Null means the metric will not be returned at all from the exporter. Zero ensures there's always at least value for comparison. Really only matters if an alert is required or how it is used in trending.

[the remark this answers is not in the document]
Comment (keithf4) [reply postgres_exporter/common/queries_nodemx.yml:51]: @cbandy just making sure this answered your question and you didn't have any further recommendations based on that

[the remark this answers is not in the document]
Comment (cbandy) [reply postgres_exporter/common/queries_nodemx.yml:51]: Works for me!

**Pull request merged**